src/error.rs

```diff
@@ -146,8 +146,190 @@ impl From<tls::Error> for Error {
 impl From<httparse::Error> for Error {
     fn from(err: httparse::Error) -> Self {
         match err {
-            httparse::Error::TooManyHeaders => Error::Capacity("Too many headers".into()),
-            e => Error::Protocol(e.to_string().into()),
+            httparse::Error::TooManyHeaders => Error::Capacity(CapacityErrorType::TooManyHeaders),
+            e => Error::Protocol(ProtocolErrorType::HttparseError(e)),
+        }
+    }
+}
+
+/// Indicates the specific type/cause of a capacity error.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum CapacityErrorType {
+    /// Too many headers provided (see [`httparse::Error::TooManyHeaders`]).
+    TooManyHeaders,
+    /// Received header is too long.
+    HeaderTooLong,
+    /// Message is bigger than the maximum allowed size.
+    MessageTooLong {
+        /// The size of the message.
+        size: usize,
+        /// The maximum allowed message size.
+        max_size: usize,
+    },
+    /// TCP buffer is full.
+    TcpBufferFull,
+}
+
+impl fmt::Display for CapacityErrorType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            CapacityErrorType::TooManyHeaders => write!(f, "Too many headers"),
+            CapacityErrorType::HeaderTooLong => write!(f, "Header too long"),
+            CapacityErrorType::MessageTooLong { size, max_size } => {
+                write!(f, "Message too long: {} > {}", size, max_size)
+            }
+            CapacityErrorType::TcpBufferFull => write!(f, "Incoming TCP buffer is full"),
+        }
+    }
+}
```

Comment:
I would slightly prefer using `thiserror` to derive this. It's distinctive advantage is that it doesn't end up in the API footprint of the library, it is just "an internal implementation helper".

Because of that, I think it would be in-scope of this PR, as it would allow us not to write (and then change) this code at all.

Reply:
Thank you for making me of `thiserror` - very handy!